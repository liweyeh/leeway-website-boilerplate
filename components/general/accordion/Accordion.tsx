/**
 * Accordion component
 * @author Carlos Romera <cromera@adasasistemas.com>
 * @author Liwei Yeh <liwei.yeh@adasasistemas.com>
 */

// Library
import React, { useState, MouseEventHandler } from 'react';

// CSS Style for this module
import styles from './accordion.module.scss';

/**
 * Interface for Accordion props
 */
interface AccordionProps {
	/** Content to be displayed in the header */
	headerContent: React.ReactNode;
	/** Content to be displayed in the body */
	children: React.ReactNode;
	/** Disable the accordion to prevent click*/
	disabled?: boolean;
	/** Custom Styles */
	customStyles?: string;
}

/**
 * Accordion generic component
 * @param {AccordionProps} Prop
 * @returns  an accordion that has a header and body and can be toggled to show/hide body
 */
export const Accordion = ({ headerContent, children, disabled, customStyles }: AccordionProps) => {
	const [active, setActive] = useState(false);

	const toggleAccordion:
		| MouseEventHandler<HTMLButtonElement | HTMLDivElement>
		| undefined = event => {
		event.stopPropagation();
		setActive(prevState => !prevState);
	};

	return (
		<div className={`${styles[`accordion`]} ${customStyles}`}>
			<div
				className={`
              ${styles[`accordion-title`]}
              ${styles[`accordion-title-${active ? 'show' : 'hidden'}`]}
            `}
				onClick={!disabled ? toggleAccordion : undefined}
			>
				<div className={styles[`accordion-title-content`]}>{headerContent}</div>
				<button
					onClick={!disabled ? toggleAccordion : undefined}
					className={styles[`accordion-title-button`]}
				>
					<div
						className={[
							styles['accordion-title-button-arrow'],
							styles[`accordion-title-button-arrow-${active ? 'show' : 'collapse'}`],
						].join(' ')}
					/>
				</button>
			</div>
			<div
				className={`
            ${styles[`accordion-body`]} 
            ${styles[`accordion-body-${active ? 'show' : 'hidden'}`]}
          `}
			>
				{children}
			</div>
		</div>
	);
};
