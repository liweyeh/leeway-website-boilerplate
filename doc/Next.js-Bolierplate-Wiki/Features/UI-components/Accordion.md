[[_TOC_]]

# Overview

- Purpose: An accordion that has a header and body and can be toggled to show/hide body
- Type: General UI
- Status: ðŸš€ **Completed**
- Code: [URL](https://dev.azure.com/ADASA-Accelerator/Website-React-Booster/_git/website-boilerplate-nextjs?path=/components/general/accordion)
- Created by: Carlos Romera <cromera@adasasistemas.com>
- Created time: 16/08/2022

# Props

\* = required

- headerContent<`React.ReactNode`>\*: Content to be displayed in the header
- children<`React.ReactNode`>\*: Content to be displayed in the body
- disabled<`boolean`>: Disable the accordion to prevent click
- customStyles<`string`>: Custom styles

# Usage

```tsx
<Accordion headerContent={<h1>2</h1>}>
	<p>2</p>
</Accordion>
```
