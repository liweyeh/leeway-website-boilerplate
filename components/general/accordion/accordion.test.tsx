import { render } from '@testing-library/react';
import { Accordion } from './Accordion';
import '@testing-library/jest-dom';

const setup = () => {
	const { asFragment } = render(
		<Accordion headerContent={<p>Title</p>}>
			<p>Body Content</p>
		</Accordion>,
	);
	return { asFragment };
};

test('test snapshot', () => {
	const { asFragment } = setup();
	expect(asFragment()).toMatchSnapshot();
});
