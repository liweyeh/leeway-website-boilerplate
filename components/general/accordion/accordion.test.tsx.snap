// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test snapshot 1`] = `
<DocumentFragment>
  <div
    class="accordion undefined"
  >
    <div
      class="
              accordion-title
              accordion-title-hidden
            "
    >
      <div
        class="accordion-title-content"
      >
        <p>
          Title
        </p>
      </div>
      <button
        class="accordion-title-button"
      >
        <div
          class="accordion-title-button-arrow accordion-title-button-arrow-collapse"
        />
      </button>
    </div>
    <div
      class="
            accordion-body 
            accordion-body-hidden
          "
    >
      <p>
        Body Content
      </p>
    </div>
  </div>
</DocumentFragment>
`;
